starboard/nplb/posix_compliance/posix_signal_test.cc

```diff
@@ -0,0 +1,337 @@
+
+#include <errno.h>
+#include <fcntl.h>
+#include <pthread.h>
+#include <signal.h>
+#include <string.h>
+#include <sys/epoll.h>
+#include <unistd.h>
+
+#include "testing/gtest/include/gtest/gtest.h"
+
+namespace starboard {
+namespace nplb {
+namespace {
+
+// --- Globals for Signal Handlers ---
+
+// A global, signal-safe flag to indicate a signal was received.
+// It must be volatile and sig_atomic_t to be safely accessed in a handler.
+volatile sig_atomic_t g_signal_received = 0;
+
+// A global file descriptor for the pipe used for reliable signal-event
+// notification. A static global is used so the C-style signal handler can
+// access it. It is set up and torn down by the test fixture.
+static int g_signal_pipe_write_fd = -1;
+
+// --- Signal Handlers ---
+
+// A simple signal handler that sets a global flag.
+void TestSignalHandler(int signum) {
+  g_signal_received = signum;
+}
+
+// An async-signal-safe handler that writes the signal number to a pipe.
+// This allows the main thread to wait on the pipe's read end, providing a
+// reliable way to detect that the handler has executed.
+void PipeWritingSignalHandler(int signum) {
+  char signal_byte = static_cast<char>(signum);
+  // This write is async-signal-safe.
+  ssize_t result = write(g_signal_pipe_write_fd, &signal_byte, 1);
+  // Avoid using library functions or complex logic in a handler.
+  // If write fails, the test waiting on the pipe will time out.
+  (void)result;
+}
+
+// --- Threading Structures and Functions ---
+
+// A structure to pass arguments to a new thread, avoiding global variables
+// for coordination.
+struct ThreadArgs {
+  pthread_mutex_t* mutex;
+  pthread_cond_t* cond;
+  bool* thread_ready;
+};
+
+// A thread entry point that uses pause() to wait for a signal.
+// It uses a condition variable to notify the main thread when it's ready.
+void* PauseThread(void* arg) {
+  ThreadArgs* thread_args = static_cast<ThreadArgs*>(arg);
+
+  // Install a handler for SIGUSR1.
+  struct sigaction sa = {};
+  sa.sa_handler = TestSignalHandler;
+  sigemptyset(&sa.sa_mask);
+  if (sigaction(SIGUSR1, &sa, NULL) != 0) {
+    return (void*)-1;  // Indicate error
+  }
+
+  // Signal the main thread that the handler is installed and we are ready.
+  pthread_mutex_lock(thread_args->mutex);
+  *thread_args->thread_ready = true;
+  pthread_cond_signal(thread_args->cond);
+  pthread_mutex_unlock(thread_args->mutex);
+
+  // Wait for a signal. pause() returns when a signal handler has returned.
+  pause();
+
+  return NULL;
+}
+
+// --- Test Fixture ---
+
+// Test fixture for POSIX signal tests. It handles saving and restoring
+// original signal dispositions and sets up synchronization primitives for
+// reliable signal and thread testing.
+class PosixSignalTest : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    // Reset the global flag before each test.
+    g_signal_received = 0;
+
+    // --- Set up pipe and epoll for reliable signal waiting ---
```

Comment:
I think the SetUp function would be easier to read if L105-120 were put into helper functions and SetUp would simply call
```
CreateNonBlockingPipe();
CreateEpollListener();
```
(feel free to change the names if you dislike those, but to that effect)